Card Component Padding System Enhancement

gigitux opened the pull request:
## What?
Closes #72175 

This PR enhances the Card component's padding system by introducing directional padding support and allowing independent padding configuration for Card, CardHeader, and CardBody components.

## Why?
The current Card component's padding system only supports predefined token sizes applied uniformly to all sides. This limits the component's flexibility in complex layouts where different padding values are needed for different sides or subcomponents.

This enhancement provides more granular control over padding while maintaining the simplicity of the existing token-based system.

## How?
1. Enhances the `size` prop to accept either a token or a directional object.
2. `CardHeader` and `CardBody` components have the `size` prop.




## Testing Instructions
1. Open the Card component storybook.
2. See the `Padding variation` story.

gigitux committed:
Enhance Card Component: Introduce directional padding options and update size type definitions

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

gigitux committed:
add changelog

gigitux committed:
add size to footer

gigitux committed:
fix changelog

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

mirka commented on packages/components/src/card/types.ts:
```diff
@@ -90,8 +98,8 @@ type MarginalSubComponentProps = BaseSubComponentProps & {
 export type FooterProps = MarginalSubComponentProps & {
 	justify?: CSSProperties[ 'justifyContent' ];
-};
+} & SizeableProps;
```
Are these additions necessary? It looks like `SizeableProps` is already included through `BaseSubComponentProps`.

gigitux replied:
Fixed with 832e1b1d988ca9efa0a7649443a096b69a9f96ed

mirka commented on packages/components/src/card/get-padding-by-size.ts:
```diff
@@ -0,0 +1,46 @@
+ * Internal dependencies
+ */
+import type { Props, SizeToken } from './types';
+import { cardPaddings } from './styles';
```
Should we move the `cardPaddings` stuff directly into this file? They're not used anywhere else. While you're at it, could you also move all the card padding values out of the `config-values` file? We want to get rid of that file eventually. I think this `get-padding-by-size.ts` file can be the canonical source of card padding sizes.

gigitux replied:
Fixed with f5d8c4738b1a01ea37f706dcddcb288e5c00a10e

mirka commented on packages/components/src/card/stories/index.story.tsx:
```diff
@@ -91,3 +91,72 @@ export const FullBleedContent: StoryObj< typeof Card > = {
 	},
 };
+
+/**
+ * The Card component supports three approaches to padding:
+ * 1. Default padding (medium) - no size prop needed
+ * 2. Token-based padding - using size tokens: xSmall (8px), small (16px), medium (24px), large (32px)
+ * 3. Directional padding - customize each side independently
+ *
+ * Each component (Card, CardHeader, CardBody) can have its own padding configuration.
+ */
```
Could we remove the exact pixel values from this story description and from the code comments? We don't want to commit to pixel values in documentation, since they could be changed at some point.

gigitux replied:
Fixed with c4af85659960dd3edbdc67b796be65cf1b348067

mirka reviewed (commented):
Looking good, thanks for the enhancement.

Could you also update the readme files? The `Card` component family isn't hooked up to our readme auto-generation system yet, sorry.

gigitux committed:
update types

gigitux committed:
move padding style in get-padding-by-size file

gigitux committed:
update story documentation

gigitux committed:
update documentation

gigitux commented:
> Looking good, thanks for the enhancement.
> 
> Could you also update the readme files? The `Card` component family isn't hooked up to our readme auto-generation system yet, sorry.

Thanks for the review! I addressed your feedback. Also, I updated the readme too: 9768cee572558c4963aa3fbe066aa179a0c1e8a6

aduth commented on packages/components/src/card/card-body/README.md:
```diff
@@ -32,10 +32,21 @@ Renders with a light gray background color.
-- Allowed values: `xSmall`, `small`, `medium`, `large`
+- Allowed values:
+-   Single size token: `xSmall`, `small`, `medium`, `large`
+-   Directional object:
```
Should we be using [logical property directions](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_logical_properties_and_values) instead, or in addition? (e.g. `blockStart`, `blockEnd`, `inlineStart`, `inlineEnd`). Are the use-case for this kind of padding customization connected to the flow of text?

gigitux replied:
> Are the use-case for this kind of padding customization connected to the flow of text?

No, it is not related to the flow of text. We need it to customize the padding of the Card elements to create a good DataForm card layout without a border. For more details, see https://github.com/WordPress/gutenberg/pull/72514

mirka replied:
Thanks for raising this point, I forgot about it. In that case, for future-proofing, it may be safest to start with just `blockStart`, `blockEnd`, and `inline` instead of top right bottom left 🤔 (Do you think it's worth having both `inlineStart` and `inlineEnd`? I don't foresee anyone having asymmetrical paddings…)

The context here is that when migrating to CSS modules we're trying to adopt a more modern system where we don't have to auto-generate a separate RTL stylesheet at compile time (i.e. everything uses logical properties or direction selectors). So if we were to add top-right-bottom-left APIs now, we would be baking in an element of ambiguity where it's unclear whether those will be auto-RTL'ed or not.

Thoughts?

gigitux replied:
Thanks for the context! Happy to implement the update. I’d prefer using inlineStart and inlineEnd to maintain consistency, but it is not a strong opinion.

gigitux replied:
I implemented it with 4a6144d02d3b99b0fd76142aca7a7d6459b4329f

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

gigitux committed:
address feedback

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

mirka commented on packages/components/src/card/card/README.md:
```diff
@@ -104,6 +115,40 @@ const Example = () => (
+- `blockStart` - Maps to `top` in horizontal writing modes
+- `blockEnd` - Maps to `bottom` in horizontal writing modes
+- `inlineStart` - Maps to `left` in horizontal writing modes
+- `inlineEnd` - Maps to `right` in horizontal writing modes
```
```suggestion
- `inlineStart` - Maps to `left` in horizontal left-to-right writing modes
- `inlineEnd` - Maps to `right` in horizontal left-to-right writing modes
```

Something like that?

mirka commented on packages/components/src/card/get-padding-by-size.ts:
```diff
@@ -0,0 +1,77 @@
+		const left = size.inlineStart;
+		const right = size.inlineEnd;
+
+		return css`
+			padding: ${ getSinglePaddingValue( top ) }
+				${ getSinglePaddingValue( right ) }
+				${ getSinglePaddingValue( bottom ) }
+				${ getSinglePaddingValue( left ) };
+		`;
```
The `padding` shorthand property is not a logical one, so it's not going to respect our logical values. We have to set each property separately (`padding-block-start`, etc).

gigitux replied:
Sorry! I completely forgot to update it! I fixed it!

mirka commented on packages/components/src/card/get-padding-by-size.ts:
```diff
@@ -0,0 +1,77 @@
+			return space( 6 );
+		case 'large':
+			return space( 8 );
+		default:
+			return '0';
+	}
```
Not sure what I think about the "fall back to 0" behavior, given that 0 is not an explicitly allowed value and the object shape marks all properties as optional. I can see it becoming a weird loophole with consumers who want to set a 0 padding.

```ts
	| {
			blockStart?: SizeToken;
			blockEnd?: SizeToken;
			inlineStart?: SizeToken;
			inlineEnd?: SizeToken;
	  };
```

I can think of two ways to address this:

1. Mark all sides as required in the object case, and fall back to `medium` for undefined or invalid sides.
2. Officially support `none` as a size token.

No strong option here — 2 is probably fine? I'll also note that there's already a `CardMedia` subcomponent that can be used for full-bleed content.

gigitux replied:
> Mark all sides as required in the object case, and fall back to medium for undefined or invalid sides.

I took this approach given that it is simple.

mirka reviewed (commented):
I'll be out next week, so I'll leave it to my pals on @WordPress/gutenberg-components for any follow up 🙏

gigitux committed:
Update packages/components/src/card/card/README.md

Co-authored-by: Lena Morita <lena@jaguchi.com>

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

gigitux committed:
improve documentation

gigitux committed:
improve logic

gigitux committed:
Merge branch '72175-card-support-more-padding-options' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

gigitux committed:
implement none

aduth commented on packages/components/src/card/get-padding-by-size.ts:
```diff
@@ -0,0 +1,82 @@
+			return space( 2 );
+		case 'small':
+			return space( 4 );
+		case 'medium':
+			return space( 6 );
```
One quirk of this approach is that if I specify `size={ { blockStart: 'medium' } }`, I'd get `space( 6 )` for the top padding, but if I specify `size="medium"`, I get `space( 4 )` for the top padding.

We could maybe have some convoluted logic for mapping sizes based on vertical or horizontal properties, but I do quite like keeping this simpler with each step mapping to `2`, `4`, `6`, and `8`.

gigitux replied:
Just to clarify, do you expect some changes here, or is it just a note?

aduth replied:
Not necessarily expecting any changes, just flagging a potential point of confusion or inconsistency of expectations. Unless you have concerns about it, I'm personally happy to try this as-is.

gigitux replied:
I agree that it could cause some confusion, but I’m not sure how to fix it without introducing breaking changes.

The idea is that with `size={{ blockStart: 'medium',  blockEnd: 'medium', inlineStart: 'medium', inlineEnd: 'medium' }}`, we can define consistent padding across all properties.

aduth commented on packages/components/src/card/get-padding-by-size.ts:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	// Handle object-based sizes
+	if ( size && typeof size === 'object' ) {
```
Is the `typeof` here check redundant? I can appreciate the explicitness, and only mention it because checking object-ness is pretty fraught (see also `isPlainObject` in various utility libraries), e.g. `typeof null === 'object'`

gigitux replied:
Addressed with 767da772f9b4869b5c20d8175341b589830d3449

aduth commented on packages/components/src/card/get-padding-by-size.ts:
```diff
@@ -0,0 +1,82 @@
+
+	// Handle object-based sizes
+	if ( size && typeof size === 'object' ) {
+		const top = size.blockStart;
```
Nit: I'd probably avoid the `top`, etc. naming because the idea of logical properties is to avoid associating with specific directions like this, even if it's more familiar due to historical usage.

gigitux replied:
Addressed with 767da772f9b4869b5c20d8175341b589830d3449

aduth commented on packages/components/src/card/get-padding-by-size.ts:
```diff
@@ -0,0 +1,82 @@
+	// The `extraSmall` size is not officially documented, but the following styles
+	// are kept for legacy reasons to support older values of the `size` prop.
+	extraSmall: xSmallCardPadding,
+};
```
`cardPaddings` is already essentially acting like a constant, so internally referencing `CONFIG` seems like unnecessary indirection over just inlining the values.


```suggestion
const xSmallCardPadding = css`
	padding: ${ space( 2 ) };
`;

export const cardPaddings = {
	none: css`
		padding: 0;
	`,
	large: css`
		padding: ${ space( 6 ) } ${ space( 8 ) };
	`,
	medium: css`
		padding: ${ space( 4 ) } ${ space( 6 ) };
	`,
	small: css`
		padding: ${ space( 4 ) } ${ space( 4 ) };
	`,
	xSmall: xSmallCardPadding,
	// The `extraSmall` size is not officially documented, but the following styles
	// are kept for legacy reasons to support older values of the `size` prop.
	extraSmall: xSmallCardPadding,
};
```

gigitux replied:
Fixed with 5c873be15d550f3f51bfa64aa8a6c93746b90bb3

aduth reviewed (approved):
Added a few minor comments, but this looks good 👍 Of my comments, the most impactful is the one about differences between horizontal and vertical spacing with values like `'medium'`.

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into 72175-card-support-more-padding-options

gigitux committed:
clean up code

gigitux committed:
remove cardPaddings object

gigitux committed:
fix changelog
